### DISCUSSION TITLE
How to connect to a Sql Server named instance on Windows

### DISCUSSION THREAD

**Discussion by Alwigi** [prisma/prisma#4548] [Q&A]
I had much trouble figuring this out so have posted here for anyone who needs it.

1. Ensure that SQL is configured to allow TCP/IP connections and that the SQL browser service is running.

2. Find the port that your instance is listening on, the SQL configuration manager may show 1433 (the default) but the actual port used by the instance may be different (mine was 50119). This query will return the port you need:

    USE master
    GO
    xp_readerrorlog 0, 1, "Server is listening on", "any", NULL, NULL, "asc"
    GO

3. To verify your tcp connection details use sqlcmd:

`sqlcmd -S tcp:Server\InstanceName,1234 -U user -P password`

4. Finally, setup your db url in schema.prisma. 
I received error 'Could not resolve SQL browser instance InstanceName' when including the full host path (Server\InstanceName) but removing the instance name resolved this.
I also had TLS errors initially (hence the 'encrypt' attribute), I guess this have to be resolved properly for production systems. 
My final (working) url looks like this:

`DATABASE_URL="sqlserver://Server:1234;initial catalog=TestDB;user=user;password=pwd;encrypt=DANGER_PLAINTEXT"`

**Comment by pimeys** [comment]
Do you use SQL Browser (regarding the InstanceName)?. Could you try:

```
DATABASE_URL="sqlserver://Server\InstanceName:1234;initial catalog=TestDB;user=user;password=pwd;encrypt=DANGER_PLAINTEXT"
```

We do run our test suite against the Windows SQL Server using the SQL Browser.

**Reply by pimeys** [reply]
If it's any help for you, you could maybe see how we set up our SQL Server in our CI:

https://github.com/prisma/tiberius/blob/master/appveyor.ps1

**Reply by pimeys** [reply]
So, how this works is the SQL Server has an UDP port open in `1234`. We then ask with `InstanceName` where is the database we want to connect, the server gives us the correct port and we do a TCP connection to that instance.

I would also suggest trying out with Microsoft's Docker version of SQL Server, might be easier to set up for development.

**Reply by Alwigi** [reply]
> Do you use SQL Browser (regarding the InstanceName)?. Could you try:
> 
> ```
> DATABASE_URL="sqlserver://Server\InstanceName:1234;initial catalog=TestDB;user=user;password=pwd;encrypt=DANGER_PLAINTEXT"
> ```
> 
> We do run our test suite against the Windows SQL Server using the SQL Browser.

I'm not sure what you're saying but I posted here to show how I got the connection to work. Putting \InstanceName in the url seems to cause trouble.

**Comment by belyaev-dev** [comment]
Many thanks! Had troubles when connecting to remote sqlserver with named instance, removing InstanceName from DATABASE_URL solved the problem.
